File: bridges/punchd-bridge/gateway/src/auth/oidc.ts
/**
 * Server-side OIDC flow for TideCloak authentication.
 *
 * Handles authorization URL construction, code exchange, token refresh,
 * and logout URL construction.
 */

import { randomBytes } from "crypto";
import { request as httpRequest } from "http";
import { request as httpsRequest } from "https";
import type { TidecloakConfig } from "../config.js";

export interface OidcEndpoints {
  authorization: string;
  token: string;
  logout: string;
}

export interface TokenResponse {
  access_token: string;
  refresh_token?: string;
  expires_in: number;
  refresh_expires_in?: number;
  token_type: string;
}

/**
 * Derive OIDC endpoints from TideCloak config.
 * @param baseUrlOverride — Optional public URL for browser-facing endpoints
 *   (e.g. ngrok URL). Server-side endpoints always use config auth-server-url.
 */
export function getOidcEndpoints(
  config: TidecloakConfig,
  baseUrlOverride?: string
): OidcEndpoints {
  const base = (baseUrlOverride ?? config["auth-server-url"]).replace(
    /\/$/,
    ""
  );
  const realmPath = `${base}/realms/${config.realm}/protocol/openid-connect`;

  return {
    authorization: `${realmPath}/auth`,
    token: `${realmPath}/token`,
    logout: `${realmPath}/logout`,
  };
}

/**
 * Build the authorization redirect URL.
 * State encodes the original URL the user was trying to access.
 */
export function buildAuthUrl(
  endpoints: OidcEndpoints,
  clientId: string,
  redirectUri: string,
  originalUrl: string
): { url: string; state: string } {
  const state = Buffer.from(
    JSON.stringify({
      nonce: randomBytes(16).toString("hex"),
      redirect: originalUrl || "/",
    })
  ).toString("base64url");

  const params = new URLSearchParams({
    client_id: clientId,
    redirect_uri: redirectUri,
    response_type: "code",
    scope: "openid",
    state,
  });

  return {
    url: `${endpoints.authorization}?${params}`,
    state,
  };
}

/**
 * Exchange authorization code for tokens.
 */
export async function exchangeCode(
  endpoints: OidcEndpoints,
  clientId: string,
  code: string,
  redirectUri: string
): Promise<TokenResponse> {
  const body = new URLSearchParams({
    grant_type: "authorization_code",
    client_id: clientId,
    code,
    redirect_uri: redirectUri,
  });

  return postTokenRequest(endpoints.token, body, "Token exchange");
}

/**
 * Refresh an access token using a refresh token.
 */
export async function refreshAccessToken(
  endpoints: OidcEndpoints,
  clientId: string,
  refreshToken: string
): Promise<TokenResponse> {
  const body = new URLSearchParams({
    grant_type: "refresh_token",
    client_id: clientId,
    refresh_token: refreshToken,
  });

  return postTokenRequest(endpoints.token, body, "Token refresh");
}

/**
 * POST to a token endpoint using http.request (not fetch).
 *
 * Uses http.request instead of fetch for reliability — Node.js fetch
 * (undici) can have DNS resolution issues in some environments.
 * The hostname is kept as-is (not replaced with 127.0.0.1) because
 * on WSL2 with Docker Desktop, "localhost" routes through a special
 * forwarding layer that 127.0.0.1 bypasses.
 */
function postTokenRequest(
  tokenUrl: string,
  params: URLSearchParams,
  label: string
): Promise<TokenResponse> {
  return new Promise((resolve, reject) => {
    const url = new URL(tokenUrl);
    const isHttps = url.protocol === "https:";
    const makeReq = isHttps ? httpsRequest : httpRequest;
    const postBody = params.toString();

    const req = makeReq(
      {
        hostname: url.hostname,
        port: url.port || (isHttps ? 443 : 80),
        path: url.pathname + url.search,
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          "Content-Length": Buffer.byteLength(postBody).toString(),
        },
      },
      (res) => {
        const chunks: Buffer[] = [];
        res.on("data", (chunk: Buffer) => chunks.push(chunk));
        res.on("end", () => {
          const text = Buffer.concat(chunks).toString("utf-8");
          if (!res.statusCode || res.statusCode >= 400) {
            console.error(`[OIDC] ${label} failed (${res.statusCode}): ${text}`);
            reject(new Error(`${label} failed (${res.statusCode}): ${text}`));
            return;
          }
          try {
            resolve(JSON.parse(text) as TokenResponse);
          } catch (e) {
            console.error(`[OIDC] ${label} response not JSON: ${text.slice(0, 200)}`);
            reject(new Error(`${label} response not JSON`));
          }
        });
      }
    );

    req.on("error", (err) => {
      console.error(`[OIDC] ${label} network error: ${err.message}`);
      reject(new Error(`${label} network error: ${err.message}`));
    });

    req.end(postBody);
  });
}

/**
 * Build TideCloak logout URL.
 */
export function buildLogoutUrl(
  endpoints: OidcEndpoints,
  clientId: string,
  postLogoutRedirectUri: string
): string {
  const params = new URLSearchParams({
    client_id: clientId,
    post_logout_redirect_uri: postLogoutRedirectUri,
  });

  return `${endpoints.logout}?${params}`;
}

/**
 * Parse the state parameter from the callback.
 */
export function parseState(state: string): { nonce: string; redirect: string } {
  try {
    return JSON.parse(Buffer.from(state, "base64url").toString("utf-8"));
  } catch {
    return { nonce: "", redirect: "/" };
  }
}

File: bridges/punchd-bridge-rs/src/auth/oidc.rs
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use serde::Deserialize;

use crate::config::TidecloakConfig;

#[derive(Clone, Debug)]
pub struct OidcEndpoints {
    pub authorization: String,
    pub token: String,
    pub logout: String,
}

#[derive(Deserialize, Debug)]
pub struct TokenResponse {
    pub access_token: String,
    pub refresh_token: Option<String>,
    pub expires_in: u64,
    pub refresh_expires_in: Option<u64>,
    pub token_type: String,
}

pub fn get_oidc_endpoints(config: &TidecloakConfig, base_url_override: Option<&str>) -> OidcEndpoints {
    let base = base_url_override
        .unwrap_or(&config.auth_server_url)
        .trim_end_matches('/');
    let realm_path = format!("{base}/realms/{}/protocol/openid-connect", config.realm);

    OidcEndpoints {
        authorization: format!("{realm_path}/auth"),
        token: format!("{realm_path}/token"),
        logout: format!("{realm_path}/logout"),
    }
}

pub fn build_auth_url(
    endpoints: &OidcEndpoints,
    client_id: &str,
    redirect_uri: &str,
    original_url: &str,
) -> (String, String) {
    let nonce = hex::encode(&random_bytes());
    let state_json = serde_json::json!({
        "nonce": nonce,
        "redirect": if original_url.is_empty() { "/" } else { original_url },
    });
    let state = URL_SAFE_NO_PAD.encode(state_json.to_string().as_bytes());

    let params = url::form_urlencoded::Serializer::new(String::new())
        .append_pair("client_id", client_id)
        .append_pair("redirect_uri", redirect_uri)
        .append_pair("response_type", "code")
        .append_pair("scope", "openid")
        .append_pair("state", &state)
        .finish();

    (format!("{}?{params}", endpoints.authorization), state)
}

pub async fn exchange_code(
    endpoints: &OidcEndpoints,
    client_id: &str,
    code: &str,
    redirect_uri: &str,
) -> Result<TokenResponse, String> {
    let body = url::form_urlencoded::Serializer::new(String::new())
        .append_pair("grant_type", "authorization_code")
        .append_pair("client_id", client_id)
        .append_pair("code", code)
        .append_pair("redirect_uri", redirect_uri)
        .finish();

    post_token_request(&endpoints.token, &body, "Token exchange").await
}

pub async fn refresh_access_token(
    endpoints: &OidcEndpoints,
    client_id: &str,
    refresh_token: &str,
) -> Result<TokenResponse, String> {
    let body = url::form_urlencoded::Serializer::new(String::new())
        .append_pair("grant_type", "refresh_token")
        .append_pair("client_id", client_id)
        .append_pair("refresh_token", refresh_token)
        .finish();

    post_token_request(&endpoints.token, &body, "Token refresh").await
}

async fn post_token_request(
    token_url: &str,
    body: &str,
    label: &str,
) -> Result<TokenResponse, String> {
    let client = reqwest::Client::builder()
        .danger_accept_invalid_certs(true)
        .build()
        .map_err(|e| format!("{label} client error: {e}"))?;

    let resp = client
        .post(token_url)
        .header("Content-Type", "application/x-www-form-urlencoded")
        .body(body.to_string())
        .send()
        .await
        .map_err(|e| format!("{label} network error: {e}"))?;

    let status = resp.status();
    let text = resp
        .text()
        .await
        .map_err(|e| format!("{label} read error: {e}"))?;

    if !status.is_success() {
        tracing::error!("{label} failed ({status}): {text}");
        return Err(format!("{label} failed ({status}): {text}"));
    }

    serde_json::from_str(&text).map_err(|e| {
        tracing::error!("{label} response not JSON: {}", &text[..200.min(text.len())]);
        format!("{label} response not JSON: {e}")
    })
}

pub fn build_logout_url(
    endpoints: &OidcEndpoints,
    client_id: &str,
    post_logout_redirect_uri: &str,
) -> String {
    let params = url::form_urlencoded::Serializer::new(String::new())
        .append_pair("client_id", client_id)
        .append_pair("post_logout_redirect_uri", post_logout_redirect_uri)
        .finish();
    format!("{}?{params}", endpoints.logout)
}

pub fn parse_state(state: &str) -> (String, String) {
    let bytes = URL_SAFE_NO_PAD.decode(state).unwrap_or_default();
    if let Ok(val) = serde_json::from_slice::<serde_json::Value>(&bytes) {
        let nonce = val["nonce"].as_str().unwrap_or("").to_string();
        let redirect = val["redirect"].as_str().unwrap_or("/").to_string();
        (nonce, redirect)
    } else {
        (String::new(), "/".to_string())
    }
}

// hex helper
mod hex {
    pub fn encode(bytes: &[u8]) -> String {
        bytes.iter().map(|b| format!("{b:02x}")).collect()
    }
}

use rand::Rng;

fn random_bytes() -> [u8; 16] {
    rand::rng().random::<[u8; 16]>()
}
